### DISCUSSION TITLE
Updating a ruleset without updating tflint causes trouble

### DISCUSSION THREAD

**Discussion by sblask** [terraform-linters/tflint#1522] [Q&A] [answered]
## Introduction

Generally speaking it would be great if there would be a better update path. With tflint 0.40.0 and tflint-ruleset-aws 0.17.0 there is no way of updating one first and then the other. So if used in different repositories, both of them have to be updated at the same for all the repositories.

I had a quite specific problem though, without having updated tflint I tried to run `tflint --init` in a repository that had the ruleset updated. I got this error:

```
Failed to initialize plugins; TFLint is not compatible with this version of the "aws" plugin. A newer TFLint or plugin version may be required. Plugin version: 10, Client versions: [11]
```

Fair enough, so I updated my tflint, ran `tflint` and got:

```
Failed to initialize plugins; Unrecognized remote plugin message:

This usually means that the plugin is either invalid or simply
needs to be recompiled to support the latest protocol.
```

But `tflint --init` gave me this:

```
Plugin `aws` is already installed
```

I found the following:

```
~/.tflint.d/plugins/github.com/terraform-linters/tflint-ruleset-aws/0.15.0 $ ./tflint-ruleset-aws
This binary is a plugin. These are not meant to be executed directly.
Please execute the program that consumes these plugins, which will
load any plugins automatically
~/.tflint.d/plugins/github.com/terraform-linters/tflint-ruleset-aws $ ./0.17.0/tflint-ruleset-aws
zsh: killed     ./0.17.0/tflint-ruleset-aws
```

Removed the 0.17.0 folder and ran `tflint --init` again and everything started working:

```
~/.tflint.d/plugins/github.com/terraform-linters/tflint-ruleset-aws $ ./0.17.0/tflint-ruleset-aws
This binary is a plugin. These are not meant to be executed directly.
Please execute the program that consumes these plugins, which will
load any plugins automatically
```

So apparently the ruleset got installed incorrectly in some way, `tflint --init` did not fix it and there is no other command to fix it.

## Additional Context

```console
$ tflint -v
TFLint version 0.40.1
$ terraform --version
Terraform v1.3.0
on darwin_arm64
```

**Comment by bendrucker** [comment, marked as answer]
> Generally speaking it would be great if there would be a better update path

Sure, but this isn't actionable at the moment. Any plugin-based system has to contend with this tradeoff of compatibility versus evolution. Terraform did this most notably from 0.11 to 0.12 with its provider plugin protocol. HashiCorp has paying customers and a full team dedicated to Terraform core, which allows them to invest much more effort up front and in long term planning. As a volunteer project, TFLint makes different tradeoffs and regularly breaks compatibility (thus 0.x) in order to allow the project to continue to rapidly evolve.

> I had a quite specific problem though

We'd need a fully reproducible example to discuss further and open an issue. Tried this quickly in a container and failed to reproduce the reported behavior. Started on tflint@0.39 and aws@0.15. Ran init. Installed `bash curl unzip` and used the installer script to update. Updated the config to use aws@0.17. Ran init. All working as expected at the end:

```sh
$ tflint --version
TFLint version 0.40.1
+ ruleset.terraform (0.1.1-bundled)
+ ruleset.aws (0.17.0)
```

**Reply by sblask** [reply]
Just tried this again and couldn't reproduce either. The ruleset file must have been corrupted for some reason. Sorry about that.

**Reply by bendrucker** [reply]
No worries, there have been some occasional hard-to-pin-down bugs around plugin downloads/extractions. All kinds of potential variables in the environment. In CI, you might even cache the plugin directory and have some other program restore that cache archive into the plugin dir rather than TFLint itself.

If you do manage to reproduce, that's definitely an issue we can investigate.
